### PR TITLE
CB-1141 Add the ability of the FMS to provision multiple FreeIPA nodes

### PR DESCRIPTION
**Pull request by jamisonbennett**
This adds support to the FreeIPA management service to support
provisioning multiple instances, but this does not change the
environment service which is hardcoded to 1 instance.

The primary gateway is chosen as the CA master, the CRL master, the
node for configuring the FreeIPA servers, and the first FreeIPA node
to be installed.

All the other FreeIPA nodes are installed next as a part of the
FreeIPA install step. During the FreeIPA post install, the network
topology is updated to be a more complete mesh with a maximum of 4
replication agreements.

This was tested manually by setting the MASTER_NODE_COUNT in
FreeIpaCreationHandler.java (which is not included in this change).

### CODE REVIEW

**Comment by keyki** [comment]
Thank you for the PR. We will review it shortly.

File: freeipa/src/main/resources/freeipa-salt/salt/freeipa/init.sls [hunk 1]
```diff
@@ -17,17 +26,36 @@ net.ipv6.conf.lo.disable_ipv6:
     - mode: 700
     - source: salt://freeipa/scripts/freeipa_install.sh
 
+{% if salt['pillar.get']('freeipa:mode') == 'FIRST_INSTALL' %}
 install-freeipa:
   cmd.run:
-    - name: /opt/salt/scripts/freeipa_install.sh && echo $(date +%Y-%m-%d:%H:%M:%S) >> /var/log/freeipa_install-executed
+    - name: /opt/salt/scripts/freeipa_install.sh "FIRST_INSTALL" && echo $(date +%Y-%m-%d:%H:%M:%S) >> /var/log/freeipa_install-executed
```
Comment (akanto) [inline freeipa/src/main/resources/freeipa-salt/salt/freeipa/init.sls]: Could you evaluate solving this problem using grains instead of pillar?

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaInstallService.java [hunk 2]
```diff
@@ -71,16 +81,36 @@ public void installFreeIpa(Long stackId) throws CloudbreakOrchestratorException
...
-                "domain", freeIpa.getDomain(),
-                "password", freeIpa.getAdminPassword());
+        Set<Object> hosts = allNodes.stream().map(n ->
+                Map.of("ip", n.getPrivateIp(),
+                        "fqdn", n.getHostname()))
+                .collect(Collectors.toSet());
+        Map<String, Object> freeipaPillar = new HashMap<>();
+        freeipaPillar.put("realm", freeIpa.getDomain().toUpperCase());
+        freeipaPillar.put("hosts", hosts);
+        freeipaPillar.put("mode", "FIRST_INSTALL");
```
Comment (akanto) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaInstallService.java]: Usually, we use grains, better to say the role field of grains, to target certain states to a particular node.  E.g.  ambari_server_standby: https://github.com/hortonworks/cloudbreak/blob/master/orchestrator-salt/src/main/resources/salt/salt/top.sls#L106

I think we can use the following roles:
freeipa_primary
freeipa_replica

File: freeipa/src/main/resources/freeipa-salt/salt/freeipa/init.sls (hunk 1 above)
Comment (jamisonbennett) [reply freeipa/src/main/resources/freeipa-salt/salt/freeipa/init.sls]: I updated it to use grains for the role assignment. I still have to see why the integration tests are failing (they don't run locally for me so I will see if I can get help on running it tomorrow).

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaInstallService.java (hunk 2 above)
Comment (jamisonbennett) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaInstallService.java]: I updated it to use grains for the role assignment. I still have to see why the integration tests are failing (they don't run locally for me so I will see if I can get help on running it tomorrow).

File: freeipa/src/main/resources/freeipa-salt/salt/freeipa/common-install.sls [hunk 3]
```diff
@@ -0,0 +1,32 @@
+/opt/salt/scripts/fix_next_update_grace_period.sh:
```
Comment (lacikaaa) [inline freeipa/src/main/resources/freeipa-salt/salt/freeipa/common-install.sls]: instead of this script, could we use a salt state `file.replace`, as I checked the script it replaces a file content

File: freeipa/src/main/resources/freeipa-salt/salt/freeipa/scripts/fix_next_update_grace_period.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+function setCaConfiguration() {
+    PROPERTY=$1
+    VALUE=$2
+    ESCAPED_PROPERTY=${PROPERTY//./[.]}
+    TMP_FILE=`mktemp`
+    cat /var/lib/pki/pki-tomcat/ca/conf/CS.cfg | sed "s/.*$ESCAPED_PROPERTY=.*/$PROPERTY=$VALUE/g" > $TMP_FILE
+    cat $TMP_FILE > /var/lib/pki/pki-tomcat/ca/conf/CS.cfg
+    rm -f $TMP_FILE
+    systemctl restart pki-tomcatd@pki-tomcat
```
Comment (lacikaaa) [inline freeipa/src/main/resources/freeipa-salt/salt/freeipa/scripts/fix_next_update_grace_period.sh]: this should be definitely be in a salt state

File: freeipa/src/main/resources/freeipa-salt/salt/freeipa/scripts/fix_next_update_grace_period.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
+
+set -e
+
+function setCaConfiguration() {
+    PROPERTY=$1
+    VALUE=$2
+    ESCAPED_PROPERTY=${PROPERTY//./[.]}
+    TMP_FILE=`mktemp`
+    cat /var/lib/pki/pki-tomcat/ca/conf/CS.cfg | sed "s/.*$ESCAPED_PROPERTY=.*/$PROPERTY=$VALUE/g" > $TMP_FILE
```
Comment (lacikaaa) [inline freeipa/src/main/resources/freeipa-salt/salt/freeipa/scripts/fix_next_update_grace_period.sh]: this could be replaced with `sed -i`, so you don't need a temp file

File: orchestrator-salt/src/main/java/com/sequenceiq/cloudbreak/orchestrator/salt/SaltOrchestrator.java [hunk 6]
```diff
@@ -492,6 +492,40 @@ public void changePrimaryGateway(GatewayConfig formerGateway, GatewayConfig newP
...
+
+            // add freeipa_replica role to the others
+            Set<String> replicaServersHostnames = allGatewayConfigs.stream()
+                    .filter(gwc -> !gwc.getHostname().equals(primaryGateway.getHostname()))
+                    .map(GatewayConfig::getHostname).collect(Collectors.toSet());
+            LOGGER.debug("Set replica FreeIPA: {}", replicaServersHostnames);
+            runSaltCommand(sc, new GrainAddRunner(replicaServersHostnames, allNodes, "freeipa_replica"), exitCriteriaModel);
+
+            // salt '*' state.highstate
+            runNewService(sc, new HighStateRunner(replicaServersHostnames, allNodes), exitCriteriaModel);
```
Comment (lacikaaa) [inline orchestrator-salt/src/main/java/com/sequenceiq/cloudbreak/orchestrator/salt/SaltOrchestrator.java]: it would be nice to see the comments as method names :)

File: orchestrator-salt/src/main/java/com/sequenceiq/cloudbreak/orchestrator/salt/SaltOrchestrator.java [hunk 7]
```diff
@@ -492,6 +492,40 @@ public void changePrimaryGateway(GatewayConfig formerGateway, GatewayConfig newP
...
+    @Override
+    public void installFreeIPA(GatewayConfig primaryGateway, List<GatewayConfig> allGatewayConfigs, Set<Node> allNodes,
+            ExitCriteriaModel exitCriteriaModel) throws CloudbreakOrchestratorException {
+        LOGGER.debug("Set primary FreeIPA: {}", primaryGateway);
+        Set<String> serverHostname = Collections.singleton(primaryGateway.getHostname());
+        try (SaltConnector sc = createSaltConnector(primaryGateway)) {
+            // set freeipa_primary role
+            runSaltCommand(sc, new GrainAddRunner(serverHostname, allNodes, "freeipa_primary"), exitCriteriaModel);
+
+            // salt '*' state.highstate
```
Comment (lacikaaa) [inline orchestrator-salt/src/main/java/com/sequenceiq/cloudbreak/orchestrator/salt/SaltOrchestrator.java]: I think `HighStateRunner` makes this comment unnecessary

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        for (TopologySegment segment : segmentsToRemove) {
+            freeIpaClient.deleteTopologySegment(topologySuffixCn, segment);
+        }
+    }
+
+    @VisibleForTesting
+    public static class UnorderedPair {
+        private final String l;
+
+        private final String r;
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java]: could you rename these to `left` and `right` as their get method states also. thanks

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+                    return s;
+                }).collect(Collectors.toSet());
+        List<TopologySuffix> topologySuffixes = freeIpaClient.findAllTopologySuffixes();
+        for (TopologySuffix topologySuffix : topologySuffixes) {
+            createMissingSegments(freeIpaClient, topologySuffix.getCn(), topology);
+            removeExtraSegements(freeIpaClient, topologySuffix.getCn(), topology);
+        }
+    }
+
+    @VisibleForTesting
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java]: if it's public, do we need this annotation?

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+                .collect(Collectors.toSet());
+        Set<TopologySegment> segmentsToRemove = freeIpaClient.findTopologySegments(topologySuffixCn).stream()
+                .filter(segment -> !topologyToKeep.contains(new UnorderedPair(segment.getLeftNode(), segment.getRightNode())))
+                .collect(Collectors.toSet());
+        for (TopologySegment segment : segmentsToRemove) {
+            freeIpaClient.deleteTopologySegment(topologySuffixCn, segment);
+        }
+    }
+
+    @VisibleForTesting
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java]: if it's public, do we need this annotation?

File: freeipa/src/main/resources/freeipa-salt/salt/freeipa/primary-install.sls [hunk 11]
```diff
@@ -0,0 +1,12 @@
+install-freeipa:
+  cmd.run:
+    - name: /opt/salt/scripts/freeipa_install.sh && echo $(date +%Y-%m-%d:%H:%M:%S) >> /var/log/freeipa_install-executed
+    - env:
+        - FPW: {{salt['pillar.get']('freeipa:password')}}
+        - DOMAIN: {{salt['pillar.get']('freeipa:domain')}}
+        - REALM: {{salt['pillar.get']('freeipa:realm')}}
+        - ADMIN_USER: {{salt['pillar.get']('freeipa:admin_user')}}
+        - FREEIPA_TO_REPLICATE: {{salt['pillar.get']('freeipa:freeipa_to_replicate')}}
```
Comment (lacikaaa) [inline freeipa/src/main/resources/freeipa-salt/salt/freeipa/primary-install.sls]: this isn't used in the script I think, could you remove? thanks

**Comment by lacikaaa** [comment]
`This was tested manually by setting the MASTER_NODE_COUNT in
FreeIpaCreationHandler.java (which is not included in this change).`
I'm wondering if it would make sense to not increase the master nodes count, and instead add slaves (https://github.com/hortonworks/cloudbreak/blob/master/freeipa-api/src/main/java/com/sequenceiq/freeipa/api/v1/freeipa/stack/model/common/instance/InstanceGroupType.java#L5) so it would be clear that there is one primary GW node and some other GW. This won't make a real difference in the end as it would result in the same output if you check this part of the code: https://github.com/hortonworks/cloudbreak/blob/master/freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/instance/MetadataSetupService.java#L59-L71

File: freeipa/src/main/resources/freeipa-salt/salt/freeipa/common-install.sls (hunk 3 above)
Comment (jamisonbennett) [reply freeipa/src/main/resources/freeipa-salt/salt/freeipa/common-install.sls]: Thanks that works much better. I didn't know about `file.replace`.

File: freeipa/src/main/resources/freeipa-salt/salt/freeipa/scripts/fix_next_update_grace_period.sh (hunk 5 above)
Comment (jamisonbennett) [reply freeipa/src/main/resources/freeipa-salt/salt/freeipa/scripts/fix_next_update_grace_period.sh]: This entire script was replaced with the salt `file.replace`.

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java (hunk 9 above)
Comment (jamisonbennett) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java]: Thanks for catching this, I meant to make it package private, so I made that correction.

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java (hunk 10 above)
Comment (jamisonbennett) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaTopologyService.java]: Thanks for catching this, I meant to make it package private, so I made that correction.

**Comment by lacikaaa** [comment]
let me rephrase it. if you send every instance in with `MASTER` type, you let the code select on of them to be `PRIMARY_GATEWAY`, other will be simply `GATEWAY`. If you send in 1 `MASTER` and the others as `SLAVE`s, then the latter ones become `GATEWAY`. That's why I wrote it won't make a big difference, at least for now.

**Comment by holleyism** [comment]
If I understand this correctly, if we make 1 Master and the others Slaves, and the master fails, we won't have any new CRLs until the master node is recovered because the Slaves won't have salt master to be able to receive configuration changes.  Isn't this a risk that if for some reason the master takes a much longer time to recover, we're left without CRL functionality?
